safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,183 @@
+	// a nonce will be provided to serialize.
+	serialize func(nonce string) string
+}
+
+type ctxKey struct{}
+
+// Nonce retrieves the nonce from the given context. If there is no nonce stored
+// in the context, an empty string is returned.
+func Nonce(ctx context.Context) string {
+	v := ctx.Value(ctxKey{})
+	if v == nil {
+		return ""
+	}
+	return v.(string)
+}
+
+// StrictCSPBuilder can be used to build a strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// If BaseURI is an empty string the base-uri directive will be set to 'none'.
+type StrictCSPBuilder struct {
+	ReportOnly    bool
+	StrictDynamic bool
+	UnsafeEval    bool
+	BaseURI       string
+	ReportURI     string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		reportOnly: s.ReportOnly,
+		needsNonce: true,
+		serialize: func(nonce string) string {
+			var b strings.Builder
+
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			b.WriteString(nonce)
+			b.WriteString("'")
+
+			if s.StrictDynamic {
+				b.WriteString(" 'strict-dynamic'")
+			}
+			if s.UnsafeEval {
+				b.WriteString(" 'unsafe-eval'")
+			}
+
+			b.WriteString("; base-uri ")
+			if s.BaseURI == "" {
+				b.WriteString("'none'")
+			} else {
+				b.WriteString(s.BaseURI)
+			}
+
+			if s.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(s.ReportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// FramingPolicy creates a new CSP policy with frame-ancestors set to 'self'.
+//
+// TODO: allow relaxation on specific endpoints according to #77.
+func FramingPolicy(reportOnly bool, reportURI string) Policy {
+	return Policy{
+		reportOnly: reportOnly,
+		serialize: func(_ string) string {
+			var b strings.Builder
+			b.WriteString("frame-ancestors 'self'")
+
+			if reportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(reportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	Policies []Policy
+}
+
+// NewInterceptor creates an interceptor from the provided policies.
+func NewInterceptor(p ...Policy) Interceptor {
+	return Interceptor{Policies: p}
+}
+
+// Default creates a new CSP interceptor with a strict nonce-based policy and a
+// framing policy, both in enforcement mode.
+func Default(reportURI string) Interceptor {
+	return NewInterceptor(StrictCSPBuilder{ReportURI: reportURI}.Build(), FramingPolicy(false, reportURI))
+}
+
+// Before claims and sets the Content-Security-Policy header and the
+// Content-Security-Policy-Report-Only header.
+func (it Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	var csps []string
+	var reportCsps []string
+	nonce := ""
+	for _, p := range it.Policies {
+		if p.needsNonce && nonce == "" {
+			nonce = generateNonce()
+		}
```

Comment:
It's enough to create the nonce once before the `for` loop. You won't need this condition here or a `needsNonce` field in the `Policy` struct.

Reply:
I was thinking that generating a nonce is an expensive action that we only want to perform when needed. But we will almost always want a nonce.